superchain/superchain.go

```diff
@@ -61,6 +70,121 @@ type AddressList struct {
 	ProxyAdmin                        Address `json:"ProxyAdmin"`
 }
 
+// ContractImplementations represent a set of contract implementations on a given network.
+// The key in the map represents the semantic version of the contract and the value is the
+// address that the contract is deployed to.
+type ContractImplementations struct {
+	L1CrossDomainMessenger       map[string]Address `yaml:"l1_cross_domain_messenger"`
+	L1ERC721Bridge               map[string]Address `yaml:"l1_erc721_bridge"`
+	L1StandardBridge             map[string]Address `yaml:"l1_standard_bridge"`
+	L2OutputOracle               map[string]Address `yaml:"l2_output_oracle"`
+	OptimismMintableERC20Factory map[string]Address `yaml:"optimism_mintable_erc20_factory"`
+	OptimismPortal               map[string]Address `yaml:"optimism_portal"`
+	SystemConfig                 map[string]Address `yaml:"system_config"`
+}
+
+// ContractVersions represents the desired semantic version of the contracts
+// in the superchain. This currently only supports L1 contracts but could
+// represent L2 predeploys in the future.
+type ContractVersions struct {
+	L1CrossDomainMessenger       string `yaml:"l1_cross_domain_messenger"`
+	L1ERC721Bridge               string `yaml:"l1_erc721_bridge"`
+	L1StandardBridge             string `yaml:"l1_standard_bridge"`
+	L2OutputOracle               string `yaml:"l2_output_oracle"`
+	OptimismMintableERC20Factory string `yaml:"optimism_mintable_erc20_factory"`
+	OptimismPortal               string `yaml:"optimism_portal"`
+	SystemConfig                 string `yaml:"system_config"`
+}
+
+// Check will sanity check the validity of the semantic version strings
+// in the ContractVersions struct.
+func (c ContractVersions) Check() error {
+	val := reflect.ValueOf(c)
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		str, ok := field.Interface().(string)
+		if !ok {
+			return fmt.Errorf("invalid type for field %s", val.Type().Field(i).Name)
+		}
+		if str == "" {
+			return fmt.Errorf("empty version for field %s", val.Type().Field(i).Name)
+		}
+		// Prepend a "v" if it's not there already
+		if !strings.HasPrefix(str, "v") {
+			str = "v" + str
+		}
+		if !semver.IsValid(str) {
+			return fmt.Errorf("invalid semver %s for field %s", str, val.Type().Field(i).Name)
+		}
+	}
+	return nil
+}
+
+// NewContractImplementations returns a new empty ContractImplementations.
+// Use this constructor to ensure that none of struct fields are nil.
+// The path argument is relative to the embedded fs that contains the implementations.
+func NewContractImplementations(path string) (ContractImplementations, error) {
```

Comment:
Can we make this a private function, and hide the path from the public interface, and instead just use the superchain name? The internal embedded FS with configuration data should be an implementation detail, something which we may want to change later.

Reply:
Sure this makes sense, good call at not leaking that abstraction